aws/rust-runtime/aws-auth/src/refresh.rs

```diff
@@ -0,0 +1,294 @@
+        // 2. Expect implementers of AsyncCredentialLoader to set the expiration if it's
+        //    not set by its origin (such as STS).
+        // 3. Support both #1 and #2 and make it configurable which is used?
+        self.maybe_expiry()
+            .expect("credential refresh loop can only be used with expiring credentials")
+    }
+
+    fn maybe_expiry(&self) -> Option<SystemTime> {
+        self.credentials.get().map(|c| c.expiry()).flatten()
+    }
+}
+
+#[derive(Copy, Clone)]
+enum RefreshLoopState {
+    Running,
+    Canceled,
+}
+
+// TODO: Add doc comment
+pub struct CancelableRefreshLoop {
+    signal: Option<RwCell<RefreshLoopState>>,
+}
+
+impl CancelableRefreshLoop {
+    pub fn cancel(self) {
+        if let Some(signal) = self.signal {
+            signal.set(RefreshLoopState::Canceled);
+        }
+    }
+}
+
+// TODO: Add doc comment
+// TODO: Refactor to have the bulk of the logic be async runtime agnostic
+// TODO: Refactor to facilitate testing
+pub struct RefreshingCredentialProvider {
+    inner: Inner,
+    cancel_signal: Option<RwCell<RefreshLoopState>>,
+}
+
+impl RefreshingCredentialProvider {
+    pub fn new(loader: Arc<dyn AsyncCredentialLoader>) -> RefreshingCredentialProvider {
+        RefreshingCredentialProvider {
+            inner: Inner {
+                loader,
+                credentials: RwCell::new(None),
+            },
+            cancel_signal: None,
+        }
+    }
+
+    #[cfg(feature = "tokio-credentials")]
+    async fn initial_refresh(&self) -> Result<(), Error> {
+        let inner = self.inner.clone();
+        let result = inner.loader.load_credentials().await;
+        match result {
+            Ok(credentials) => self.inner.credentials.set(Some(credentials)),
+            Err(err) => return Err(Error::RefreshFailed(Box::new(err))),
+        }
+        Ok(())
+    }
+
+    #[cfg(feature = "tokio-credentials")]
+    pub async fn spawn_refresh_loop(&mut self) -> Result<CancelableRefreshLoop, Error> {
+        self.initial_refresh().await?;
+
+        // Establish cancellation communication
+        let cancel_signal = RwCell::new(RefreshLoopState::Running);
+        self.cancel_signal = Some(cancel_signal.clone());
+
+        // Start watchdog thread
+        let inner = self.inner.clone();
+        let watchdog_signal = cancel_signal.clone();
+        thread::Builder::new()
+            .stack_size(WATCHDOG_STACK_SIZE_BYTES)
+            .name("RefreshingCredentialProvider-watchdog".into())
+            .spawn(move || {
+                Self::watchdog_thread(watchdog_signal, inner);
+            })
+            .map_err(|err| Error::ThreadSpawnFailed(Box::new(err)))?;
+
+        Ok(CancelableRefreshLoop {
+            signal: Some(cancel_signal),
+        })
+    }
+
+    #[cfg(feature = "tokio-credentials")]
+    fn watchdog_thread(signal: RwCell<RefreshLoopState>, inner: Inner) {
+        while let RefreshLoopState::Running = signal.get() {
```

Comment:
You will probably want to implement the watch loop using async so you can select over the shutdown signal and the refreshing logic.